storage: missing intent in `TestMVCCHistories/range_tombstone_writes`

A `testrace` run on #87403 resulted in this test failure:

```
=== RUN   TestMVCCHistories/range_tombstone_writes
    mvcc_history_test.go:342:
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/testdata/mvcc_histories/range_tombstone_writes:443: del_range_ts k=+ end=z ts=1
        output didn't match expected:
        @@ -12,7 +12,7 @@
         data: "g"/4.000000000,0 -> /<empty>
         data: "g"/2.000000000,0 -> /BYTES/g2
         meta: "h"/0,0 -> txn={<nil>} ts=0,0 del=false klen=0 vlen=0 raw=/BYTES/inline mergeTs=<nil> txnDidNotUpdateMeta=false
         meta: "i"/0,0 -> txn={id=00000000 key=/Min pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true
         data: "i"/7.000000000,0 -> /BYTES/i7
        -error: (*assert.withAssertionFailure:) can't write MVCC range tombstone across local keyspan {\x00-z}/1.000000000,0
        +error: (*withstack.withStack:) intentIter at intent, but iter not at provisional value
    --- FAIL: TestMVCCHistories/range_tombstone_writes (0.08s)
```

The change in that PR should not have caused this failure. Instead, it indicates a problem either with the `intentInterleavingIter` or with Pebble itself, possibly a point key not being surfaced.

Jira issue: CRDB-19371

--- a/pkg/kv/kvserver/mvcc_gc_queue_test.go
+++ b/pkg/kv/kvserver/mvcc_gc_queue_test.go
@@ -482,6 +482,7 @@ func TestFullRangeDeleteHeuristic(t *testing.T) {
 	defer eng.Close()
 
 	rng, _ := randutil.NewTestRand()
+	localMax, maxKey := keys.LocalMax, keys.MaxKey
 
 	var keys []roachpb.Key
 	fillDataForStats := func(rw storage.ReadWriter, keyCount, valCount int) (enginepb.MVCCStats, hlc.Timestamp) {
@@ -505,7 +506,7 @@ func TestFullRangeDeleteHeuristic(t *testing.T) {
 
 	deleteWithTombstone := func(rw storage.ReadWriter, delTime hlc.Timestamp, ms *enginepb.MVCCStats) {
 		require.NoError(t, storage.MVCCDeleteRangeUsingTombstone(
-			ctx, rw, ms, []byte{0}, []byte{0xff}, delTime, hlc.ClockTimestamp{}, nil, nil, false, 1, nil))
+			ctx, rw, ms, localMax, maxKey, delTime, hlc.ClockTimestamp{}, nil, nil, false, 1, nil))
 	}
 	deleteWithPoints := func(rw storage.ReadWriter, delTime hlc.Timestamp, ms *enginepb.MVCCStats) {
 		for _, key := range keys {

--- a/pkg/storage/intent_interleaving_iter.go
+++ b/pkg/storage/intent_interleaving_iter.go
@@ -169,7 +169,7 @@ var intentInterleavingIterPool = sync.Pool{
 }
 
 func isLocal(k roachpb.Key) bool {
-	return len(k) == 0 || keys.IsLocal(k)
+	return k.Compare(keys.LocalMax) < 0
 }
 
 func newIntentInterleavingIterator(reader Reader, opts IterOptions) MVCCIterator {

--- a/pkg/storage/mvcc.go
+++ b/pkg/storage/mvcc.go
@@ -3187,7 +3187,7 @@ func MVCCDeleteRangeUsingTombstone(
 
 	// We currently don't allow MVCC range tombstones across the local keyspace,
 	// to be safe. This wouldn't handle MVCC stats (SysBytes) correctly either.
-	if keys.IsLocal(startKey) || keys.IsLocal(endKey) {
+	if startKey.Compare(keys.LocalMax) < 0 {
 		return errors.AssertionFailedf("can't write MVCC range tombstone across local keyspan %s",
 			rangeKey)
 	}

--- a/pkg/storage/mvcc_history_test.go
+++ b/pkg/storage/mvcc_history_test.go
@@ -1099,8 +1099,9 @@ func cmdDeleteRangeTombstone(e *evalCtx) error {
 	var msCovered *enginepb.MVCCStats
 	if cmdDeleteRangeTombstoneKnownStats && !e.hasArg("noCoveredStats") {
 		// Some tests will submit invalid MVCC range keys, where e.g. the end key is
-		// before the start key -- ignore them to avoid iterator panics.
-		if key.Compare(endKey) < 0 {
+		// before the start key -- don't attempt to compute covered stats for these
+		// to avoid iterator panics.
+		if key.Compare(endKey) < 0 && key.Compare(keys.LocalMax) >= 0 {
 			ms, err := ComputeStats(e.engine, key, endKey, ts.WallTime)
 			if err != nil {
 				return err

--- a/pkg/storage/testdata/mvcc_histories/range_tombstone_writes
+++ b/pkg/storage/testdata/mvcc_histories/range_tombstone_writes
@@ -418,7 +418,7 @@ meta: "h"/0,0 -> txn={<nil>} ts=0,0 del=false klen=0 vlen=0 raw=/BYTES/inline me
 meta: "i"/0,0 -> txn={id=00000000 key=/Min pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true
 data: "i"/7.000000000,0 -> /BYTES/i7
 
-# Writing across the local keyspace should error.
+# Writing across the local keyspace should error, as should writing from \x00.
 run error
 del_range_ts k=%a end=%b ts=1
 ----
@@ -439,3 +439,24 @@ meta: "h"/0,0 -> txn={<nil>} ts=0,0 del=false klen=0 vlen=0 raw=/BYTES/inline me
 meta: "i"/0,0 -> txn={id=00000000 key=/Min pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true
 data: "i"/7.000000000,0 -> /BYTES/i7
 error: (*assert.withAssertionFailure:) can't write MVCC range tombstone across local keyspan /Local/Range/"{a"-b"}/1.000000000,0
+
+run error
+del_range_ts k=+ end=z ts=1
+----
+>> at end:
+rangekey: {a-b}/[10.000000000,0=/<empty>]
+rangekey: {b-d}/[4.000000000,0=/<empty>]
+rangekey: {k-p}/[4.000000000,0=/<empty>]
+data: "a"/4.000000000,0 -> /<empty>
+data: "a"/2.000000000,0 -> /BYTES/a2
+data: "b"/3.000000000,0 -> /BYTES/b3
+meta: "d"/0,0 -> txn={id=00000000 key=/Min pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true
+data: "d"/7.000000000,0 -> /BYTES/d7
+data: "d"/4.000000000,0 -> /BYTES/d4
+data: "f"/4.000000000,0 -> /BYTES/f4
+data: "g"/4.000000000,0 -> /<empty>
+data: "g"/2.000000000,0 -> /BYTES/g2
+meta: "h"/0,0 -> txn={<nil>} ts=0,0 del=false klen=0 vlen=0 raw=/BYTES/inline mergeTs=<nil> txnDidNotUpdateMeta=false
+meta: "i"/0,0 -> txn={id=00000000 key=/Min pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true
+data: "i"/7.000000000,0 -> /BYTES/i7
+error: (*assert.withAssertionFailure:) can't write MVCC range tombstone across local keyspan {\x00-z}/1.000000000,0

